runtime/ghex_comm/run.cpp

```diff
@@ -46,41 +46,53 @@ runtime::runtime(int num_threads)
             }
           },
           MPI_Finalize),
-      m_num_threads(num_threads) {
+      m_num_threads(num_threads), m_cart_dims(cart_dims),
+      m_thread_cart_dims(thread_cart_dims), m_device_mapping(num_threads, 0) {
   int size, rank;
   MPI_Comm_size(MPI_COMM_WORLD, &size);
   MPI_Comm_rank(MPI_COMM_WORLD, &rank);
 
   if (size > 1 && rank != 0)
     std::cout.setstate(std::ios_base::failbit);
 
+  MPI_Dims_create(size, 2, m_cart_dims.data());
+  if (m_cart_dims[0] * m_cart_dims[1] != size) {
+    throw std::runtime_error(
+        "the product of cart dims must be equal to the number of MPI ranks.");
+  }
+  if (m_thread_cart_dims[0] * m_thread_cart_dims[1] != num_threads) {
+    if ((m_thread_cart_dims[0] != 0 && m_thread_cart_dims[1] != 0) ||
+        m_thread_cart_dims[0] > num_threads ||
+        m_thread_cart_dims[1] > num_threads)
+      throw std::runtime_error(
+          "the product of thread cart dims must be equal to the number of "
+          "threads per rank.");
+    if (m_thread_cart_dims[0] == 0 && m_thread_cart_dims[1] == 0)
+      m_thread_cart_dims =
+          partition_factors(factorize(num_threads), std::array<int, 2>{1, 1});
+    else if (m_thread_cart_dims[0] == 0)
+      m_thread_cart_dims[0] = num_threads / m_thread_cart_dims[1];
+    else
+      m_thread_cart_dims[1] = num_threads / m_thread_cart_dims[0];
+    if (m_thread_cart_dims[0] * m_thread_cart_dims[1] != num_threads)
+      throw std::runtime_error(
+          "the product of thread cart dims must be equal to the number of "
+          "threads per rank.3");
+  }
+
 #ifdef __CUDACC__
-  if (num_threads > 1) {
-    int device_count = 1;
-    if (cudaGetDeviceCount(&device_count) != cudaSuccess)
-      throw std::runtime_error("cudaGetDeviceCount failed");
-    MPI_Comm shmem_comm;
-    MPI_Comm_split_type(MPI_COMM_WORLD, MPI_COMM_TYPE_SHARED, 0, MPI_INFO_NULL,
-                        &shmem_comm);
-    int node_rank = 0;
-    MPI_Comm_rank(shmem_comm, &node_rank);
-    MPI_Comm_free(&shmem_comm);
-    const int device_id = node_rank % device_count;
-    if (cudaSetDevice(device_id) != cudaSuccess)
-      throw std::runtime_error("cudaSetDevice failed");
-    if (device_count > 1) {
-      for (int i = 0; i < device_count; ++i) {
-        if (i != device_id) {
-          int flag;
-          if (cudaDeviceCanAccessPeer(&flag, device_id, i) != cudaSuccess)
-            throw std::runtime_error("cudaDeviceAccessPeer failed");
-          if (flag) {
-            cudaDeviceEnablePeerAccess(i, 0);
-          }
-        }
-      }
-    }
+  if (device_mapping.size() > 0) {
+    if (device_mapping.size() != size * num_threads)
```

Comment:
I am a bit confused here: `size` is the size of `COMM_WORLD`, that is, total number of MPI ranks, right? But according to the help message below, this should work per-node, not globally:
```
  options("device-mapping",
          "node device mapping: device id per sub-domain in the format "
          "I_0:I_1:...:I_(N-1) "
          "where I_i are cuda device ids "
          "and N = #ranks-per-node x S",
          "M");
```

Reply:
forgot to add the node_size when refactoring :) Here it is.